### ISSUE TITLE
7卡4090训练爆显存

### ISSUE THREAD

**Issue by scut-zx** [Liuziyu77/Visual-RFT#75]
在训练Qwen2-VL-2B-Instruct的时候，即使使用zero3_offload，num_generations只能设置为2，超过2就会爆显存。Qwen2-VL-7B-Instruct即使使用zero3_offload和num_generations 2也会爆显存。想问下是优化问题还是显存不够？
具体参数如下：
> torchrun --nproc_per_node="7" \
>     --nnodes="1" \
>     --node_rank="0" \
>     --master_addr="127.0.0.1" \
>     --master_port="12345" \
>     src/virft/src/open_r1/grpo.py \
>     --output_dir ${SAVE_PATH}  \
>     --model_name_or_path ${CKPT_PATH} \
>     --dataset_name ${DATA_PATH} \
>     --deepspeed src/virft/local_scripts/zero3_offload.json \
>     --max_prompt_length 1024 \
>     --per_device_train_batch_size 1 \
>     --gradient_accumulation_steps 2 \
>     --logging_steps 1 \
>     --bf16 \
>     --report_to wandb \
>     --gradient_checkpointing false \
>     --attn_implementation flash_attention_2 \
>     --max_pixels 401408 \
>     --num_train_epochs 2 \
>     --run_name Qwen2-VL-2B_GRPO_coco_base65cate \
>     --save_steps 100 \
>     --save_only_model true \
>     --num_generations 2

**Comment by jacoblam3112** [comment]
I am also curious to know how to manage the memory. For instance, is it possible that each card can be used to generate one example and then we combine the rewards. And then on top, we could do gradient accumulation ?

**Comment by LiuRicky** [comment]
> 在训练Qwen2-VL-2B-Instruct的时候，即使使用zero3_offload，num_generations只能设置为2，超过2就会爆显存。Qwen2-VL-7B-Instruct即使使用zero3_offload和num_generations 2也会爆显存。想问下是优化问题还是显存不够？ 具体参数如下：
> 
> > torchrun --nproc_per_node="7" 
> > --nnodes="1" 
> > --node_rank="0" 
> > --master_addr="127.0.0.1" 
> > --master_port="12345" 
> > src/virft/src/open_r1/grpo.py 
> > --output_dir ${SAVE_PATH}  
> > --model_name_or_path ${CKPT_PATH} 
> > --dataset_name ${DATA_PATH} 
> > --deepspeed src/virft/local_scripts/zero3_offload.json 
> > --max_prompt_length 1024 
> > --per_device_train_batch_size 1 
> > --gradient_accumulation_steps 2 
> > --logging_steps 1 
> > --bf16 
> > --report_to wandb 
> > --gradient_checkpointing false 
> > --attn_implementation flash_attention_2 
> > --max_pixels 401408 
> > --num_train_epochs 2 
> > --run_name Qwen2-VL-2B_GRPO_coco_base65cate 
> > --save_steps 100 
> > --save_only_model true 
> > --num_generations 2

Maybe you can try the code in Seg-Zero, which is based on veRL and can split model to multiple GPU for sampling.

**Comment by ChangYu-beginner** [comment]
> I am also curious to know how to manage the memory. For instance, is it possible that each card can be used to generate one example and then we combine the rewards. And then on top, we could do gradient accumulation ?

Thank you for your advice. I will have a try. I’m not familiar with the distribution calculation, so I can't answer your question.

**Comment by ChangYu-beginner** [comment]
Setting `--gradient_checkpointing True` can effectively reduce GPU memory usage, but does this have any impact on the accuracy of model training? 
The chart below shows the results of training ViRFT_COCO_base65 using Qwen2-VL-2B-Instruct. The purple line represents the case where `--gradient_checkpointing True` was used along with `--num_generations 8`, while the cyan line represents `--gradient_checkpointing True` with `--num_generations 2`. The differences in results appear to be minimal. Does this suggest that `--num_generations` does not significantly affect the outcome?

![Image](https://github.com/user-attachments/assets/3ea7484c-c71a-47e3-a2b4-0a71725da4c9)

**Comment by jacoblam3112** [comment]
@ChangYu-beginner  checkpointing should not impact model accuracy . It's just a memory optimization technique that trades computation time for memory by recomputing intermediate activations during the backward pass instead of storing them.
By the way, may I know how long it took to train 500 steps on your hardware setup ?

**Comment by ChangYu-beginner** [comment]
> checkpointing should not impact model accuracy . It's just a memory optimization technique that trades computation time for memory by recomputing intermediate activations during the backward pass instead of storing them.
> By the way, may I know how long it took to train 500 steps on your hardware setup ?

In the case of the parameters given above and 7 * 4090 (24G), it was used for 5 hours to train 500 steps

**Comment by ai-meng** [comment]
> [@ChangYu-beginner](https://github.com/ChangYu-beginner) checkpointing should not impact model accuracy . It's just a memory optimization technique that trades computation time for memory by recomputing intermediate activations during the backward pass instead of storing them. By the way, may I know how long it took to train 500 steps on your hardware setup ?

一般训个200 steps 不就够了吗？看作者推荐

**Comment by ChangYu-beginner** [comment]
> > [@ChangYu-beginner](https://github.com/ChangYu-beginner) checkpointing should not impact model accuracy . It's just a memory optimization technique that trades computation time for memory by recomputing intermediate activations during the backward pass instead of storing them. By the way, may I know how long it took to train 500 steps on your hardware setup ?
> 
> 一般训个200 steps 不就够了吗？看作者推荐

确实，但是卡没人用就能跑久点：)

**Comment by Haiyang-Zhou** [comment]
> > > [@ChangYu初学者](https://github.com/ChangYu-beginner)检查点不应影响 Model Accuracy 。它只是一种内存优化技术，通过在向后传递期间重新计算中间激活而不是存储它们来将计算时间换成内存。顺便说一句，我可以知道在您的硬件设置上训练 500 个步骤需要多长时间吗？
> > 
> > 
> > 一般训个200 steps 不就够了吗？看作者推荐
> 
> 确实，但是卡没人用就能跑久点：)

哥，可以加个联系方式么，我有两张4090，最近刚开始学视觉大模型微调，想找您学习下

**Comment by ChangYu-beginner** [comment]
@Haiyang-Zhou 我也是最近刚入门，可能没法给你啥帮助。给你些参考资料吧：

- 新手多模态大模型学习路线，请各位分享自己的来时路？：https://www.zhihu.com/question/14222627229/answer/118635644662

- 多模态大模型入门相关：https://zhuanlan.zhihu.com/p/736744658

- deepseek-r1开源复现方法整理：https://zhuanlan.zhihu.com/p/21062322587

- 还有就是看下B站Up主，比如[五道口纳什](https://space.bilibili.com/59807853)，[chaofa用代码打点酱油](https://space.bilibili.com/12420432)的视频，学一下相关技术。

- 剩下就是看各种领域的论文了

纯属抛砖引玉，欢迎大家补充

**Comment by MING-ZCH** [comment]
> > [@ChangYu-beginner](https://github.com/ChangYu-beginner) checkpointing should not impact model accuracy . It's just a memory optimization technique that trades computation time for memory by recomputing intermediate activations during the backward pass instead of storing them. By the way, may I know how long it took to train 500 steps on your hardware setup ?
> 
> 一般训个200 steps 不就够了吗？看作者推荐

同默认 7B bash 参数设置，实测 200 steps 效果和 500 steps 差距不大

**Comment by linengcs** [comment]
> > > [@ChangYu-beginner](https://github.com/ChangYu-beginner) checkpointing should not impact model accuracy . It's just a memory optimization technique that trades computation time for memory by recomputing intermediate activations during the backward pass instead of storing them. By the way, may I know how long it took to train 500 steps on your hardware setup ?
> > 
> > 
> > 一般训个200 steps 不就够了吗？看作者推荐
> 
> 同默认 7B bash 参数设置，实测 200 steps 效果和 500 steps 差距不大

请问你跑7B用到的计算资源是多少呢，我8卡4090，开gradient checkpoint以及num generations为2也是爆显存了

**Comment by ChangYu-beginner** [comment]
> > > > [@ChangYu-beginner](https://github.com/ChangYu-beginner) checkpointing should not impact model accuracy . It's just a memory optimization technique that trades computation time for memory by recomputing intermediate activations during the backward pass instead of storing them. By the way, may I know how long it took to train 500 steps on your hardware setup ?
> > > 
> > > 
> > > 一般训个200 steps 不就够了吗？看作者推荐
> > 
> > 
> > 同默认 7B bash 参数设置，实测 200 steps 效果和 500 steps 差距不大
> 
> 请问你跑7B用到的计算资源是多少呢，我8卡4090，开gradient checkpoint以及num generations为2也是爆显存了

大概每张卡显存是16000MB这样。目前已知的降显存办法包括：

- 将 --deepspeed 设置使用zero3_offload.json
- 降低 --max_prompt_length
- 降低 --per_device_train_batch_size 以及 --gradient_accumulation_steps 2，真实的batch_size是这两者的乘积
- 设置 --gradient_checkpointing 为 True （非常实用）
- 降低 --max_pixels
- 降低 --num_generations

希望对你有帮助：）

**Comment by frozen30** [comment]
请问你试过用qwen2.5-VL-3B的训练过吗？会不会爆显存
